boot/freeldr/bootsect/ntfs.S

```diff
@@ -0,0 +1,829 @@
+#define ATTRIBUTE_INDEX_ALLOCATION HEX(A0)
+
+StartSearch:
+    // Compute MFT start sector
+    mov   eax, dword ptr [MftLocation]
+    mov   cl, byte ptr [SectsPerCluster]
+    movzx ecx, cx
+    mul   ecx
+    mov   dword ptr MFTStartSector, eax
+
+    // Compute size of MFT entry in sectors
+    mov   al, byte ptr [ClustersPerMftRecord]
+    movzx ax, al
+    test  al, HEX(80)
+    jnz   NegativeOffset
+    mov   cx, word ptr [SectsPerCluster]
+    mul   cx
+    mov   word ptr [SectsPerMFT], cx
+    jmp   SearchNext
+    NegativeOffset:
+    not   al
+    mov   cl, al
+    sub   cl, 8
+    xor   ax, ax
+    inc   ax
+    shl   ax, cl
+    mov   word ptr [SectsPerMFT], ax
+
+    SearchNext:
+    mov   bp, sp
+    sub   sp, HEX(10)
+
+    // Read Root Directory MFT
+    mov   ax, HEX(2000)
+    mov   es, ax
+    xor   bx, bx
+    mov   eax, 5
+    call  ReadInode
+
+    xor   ax, ax
+    call  ExploreIndexRoot
+
+    push  eax
+    mov   ax, HEX(A00)
+    mov   es, ax
+    pop   eax
+    call  ReadInode
+
+    xor   ax, ax
+    mov   ebx, HEX(30)
+    call  FindAttributeHdr
+    mov   si, ax
+    // Move to the attribute data
+    add   si, word ptr es:[si + HEX(14)]
+
+    // We don't support compressed, sparse or encrypted Freeldr
+    test  dword ptr es:[si + HEX(38)], HEX(4a00)
+    jnz   CompressedFreeldr
+
+    // Compute size in clusters
+    mov   eax, dword ptr es:[si + HEX(28)]
+    mov   cl, byte ptr [SectsPerCluster]
+    xor   edx, edx
+    div   ecx
+    mov   cx, word ptr [BytesPerSector]
+    movzx ecx, cx
+    xor   edx, edx
+    div   ecx
+    mov   edx, eax
+
+    push  ax
+    mov   si, offset msgLoading
+    call  PutChars
+    pop   ax
+
+    xor   ax, ax
+    mov   ebx, HEX(80)
+    call  FindAttributeHdr
+    mov   si, ax
+    add   ax, word ptr es:[si + HEX(20)]
+    xor   ecx, ecx
+    xor   bx, bx
+    push  FREELDR_BASE / 16
+    pop   fs
+    FreeLdrLoad:
+        pushad
+        call  ReadNonResidentAttribute
+        popad
+        mov   bx, fs
+        add   bx, HEX(100)
+        mov   fs, bx
+        xor   bx, bx
+        inc   ecx
+        cmp   ecx, edx
+        jbe   FreeLdrLoad
+
+    mov    dl, byte ptr [BootDrive]
+    mov    dh, byte ptr [BootPartition]
+    ljmp16 0, FREELDR_BASE
+
+CompressedFreeldr:
+    mov   si, offset msgFreeldrCompressed
+    call  PutChars
+    jmp   Reboot
+
+// Finds Freeldr.sys into the directory tree and returns the
+// inode index
+// [ES:AX] address of the MFT record
+ExploreIndexRoot:
+    #define fileRecordBase 2
+    #define fileRecord 4
+    #define indexRootData 6
+    #define allocationRunList 8
+
+    push  bp
+    mov   bp, sp
+    push  bx
+    push  si
+    push  di
+    sub   sp, HEX(10)
+    mov   word ptr [bp - fileRecordBase], es
+    mov   word ptr [bp - fileRecord], ax
+
+    mov   ebx, ATTRIBUTE_INDEX_ROOT
+    call  FindAttributeHdr
+    test  ax, ax
+    jz    near ptr PrintFileSystemError         // fail if no INDEX_ROOT
```

Comment:
The `near` can be omitted?

Reply:
MSVC build whines if not present

Reply:
This would then be our first asm code with near ptr.
Could you see whether by removing `near ptr` it works then? (that's how all of our other code does it; alternatively it uses `short`)

Reply:
For some reason MSVC stopped whining on rhis and compiles fine without.

Reply:
Maybe it's part of their new set of bugs they "introduced" in MASM when it compiles 16-bit code?
See another instance of this there: https://developercommunity.visualstudio.com/t/MASMML-x86:-Inconsistent-large-mov/1385890